### ISSUE TITLE
[Question] (what is advantage of introducing concept of expression and statement in rust?)

### ISSUE THREAD

**Issue by jessi68** [kaist-cp/cs492-concur#53]
In rust, it differentiate concept of statement and expression unlike another language.
I curious about purpose or advantage of differentiate and introduce those conecepts in rust.

**Comment by yusungsim** [comment]
In personal opinion, it is one of functional programming element in Rust.

There are many features and elements in Rust that are influenced by functional paradigm, including strict division of statement(executed in sequence) and expression (evaluates into value). Rust require both branches of if-else statement to return value of same type. It is also frequently used restriction in functional programming languages.

Then why Rust adopt functional programming elements? (Also a personal opinion) Functional programming allows programmers to forget about side effects and simply consider pure functions. This aligns well with Rust's motivation, which enables programmers to ensure high-level safety with somewhat restricted control. Ownership type system of Rust is also and the most important mechanism to provide safety AND control for programmer.

In summary, Rust adopted many FP features such as statement vs. expression, which supplements programmers to conveniently reason about high-level safety.

**Comment by tomtomjhj** [comment]
Can you give an example of a language that doesn't differentiate concept of statement and expression?

**Comment by YujinGaya** [comment]
Rust does not introduces new concepts. Most languages differentiate those two concepts. For instance in C, `1 + 2` is an expression where `if` is a statement.

**Comment by jeehoonkang** [comment]
Thanks for a productive discussion.  It's an epitome of a constructive Q&A.

Closing, but feel free to reopen it if you have further questions.
